src/python/espressomd/actors.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (C) 2010-2019 The ESPResSo project
+#
+# This file is part of ESPResSo.
+#
+# ESPResSo is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ESPResSo is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from . import highlander
+from . import script_interface
+
+
+class Actors:
+
+    """
+    Container for actor objects.
+    """
+
+    active_actors = []
+
+    def __del__(self):
+        self.clear()
+
+    def __getstate__(self):
+        return self.active_actors
+
+    def __setstate__(self, active_actors):
+        self.active_actors[:] = []
+        for actor in active_actors:
+            self.active_actors.append(actor)
+            actor._activate()
+
+    def add(self, actor):
+        """
+        Parameters
+        ----------
+        actor :
+            Actor to add to this container.
+
+        """
+        if actor in Actors.active_actors:
+            raise highlander.ThereCanOnlyBeOne(actor)
+
+        if isinstance(actor, script_interface.ScriptInterfaceHelper):
+            actor._activate()
+
+        self.active_actors.append(actor)
+
+        if not isinstance(actor, script_interface.ScriptInterfaceHelper):
+            actor._activate()
```

Comment:
where does it come from, that for `ScriptInterfaceHelper`-objects we need to call `_activate()` before appending it to the list?

Reply:
Here is the intended workflow:
- `_activate()` calls the C++ actor registration code that adds the actor in the global optional variant
- the now active actor in the core runs the sanity checks and tuning algorithm (if the method supports tuning)
- if checks or tuning fail, a C++ exception is thrown, which is handled in the [registration code](https://github.com/espressomd/espresso/pull/4506/files/0d6f2c1ba994eeac24eb41c0b11581d668811b07#diff-f9f14648861a7b0f5fa3655eedb94ffb693693d1e0410be86582aae1cdb47d38R54-R62)
   - the actor is safely removed from the core global
   - the exception is re-thrown
   - on MPI rank 0, the C++ exception becomes a Python exception
   - line 54 in actors.py throws
   - the python object is never added to the list of active actors

Since removing an actor from the actor list will trigger sanity checks in the core, we cannot remove a python actor if it is not also in the core. We also cannot keep an invalid actor in the core. LB and EK actors still rely on the old workflow.